picoquic/sender.c

```diff
@@ -1011,36 +1026,39 @@ static void picoquic_cnx_set_next_wake_time_init(picoquic_cnx_t* cnx, uint64_t c
             next_time = path_x->next_pacing_time;
         }
         else {
-            /* Consider delayed ACK */
-            if (cnx->ack_needed) {
-                next_time = cnx->highest_ack_time + cnx->ack_delay_local;
-            }
-
-            /* Consider path challenges */
-            if (path_x->challenge_verified == 0) {
-                uint64_t next_challenge_time = path_x->challenge_time + path_x->retransmit_timer;
-                if (current_time < next_challenge_time) {
-                    if (next_time > next_challenge_time) {
-                        next_time = next_challenge_time;
-                    }
-                } else {
-                    next_time = current_time;
+            for (picoquic_packet_context_enum pc = 0; pc < picoquic_nb_packet_context; pc++) {
+                picoquic_packet* p = cnx->pkt_ctx[pc].retransmit_oldest;
+                /* Consider delayed ACK */
+                if (cnx->pkt_ctx[pc].ack_needed) {
+                    next_time = cnx->pkt_ctx[pc].highest_ack_time + cnx->pkt_ctx[pc].ack_delay_local;
                 }
-            }
 
-            if (p != NULL) {
-                if (cnx->nb_retransmit == 0) {
-                    if (p->send_time + path_x->retransmit_timer < next_time) {
-                        next_time = p->send_time + path_x->retransmit_timer;
+                if (p != NULL) {
+                    if (cnx->pkt_ctx[pc].nb_retransmit == 0) {
+                        if (p->send_time + path_x->retransmit_timer < next_time) {
+                            next_time = p->send_time + path_x->retransmit_timer;
+                        }
                     }
-                }
-                else {
-                    if (p->send_time + (1000000ull << (cnx->nb_retransmit - 1)) < next_time) {
-                        next_time = p->send_time + (1000000ull << (cnx->nb_retransmit - 1));
+                    else {
+                        if (p->send_time + (1000000ull << (cnx->pkt_ctx[pc].nb_retransmit - 1)) < next_time) {
+                            next_time = p->send_time + (1000000ull << (cnx->pkt_ctx[pc].nb_retransmit - 1));
+                        }
                     }
                 }
             }
+        }
+    }
 
+    /* Consider path challenges */
+    if (path_x->challenge_verified == 0) {
+        uint64_t next_challenge_time = path_x->challenge_time + path_x->retransmit_timer;
```

Comment:
This is perfectly fine, though I find nesting levels slow me down, when I am reading code.  I would had prefered the same as below.  This makes code shorter, fewer nesting levels, comparisons are always "left less than right".  Being more straightforward, I bet that if there were a logical mistake in this, it would become more immediately apparent.

```
if (next_challenge_time <= current_time) {
    next_time = current_time;
} else if (next_challenge_time < next_time) {
    next_time = next_challenge_time;
}
```

Reply:
Yes. This is not new code, but code moved from the previous "single context" implementation. I wanted to keep changes to a minimum -- debug was hard enough...

Reply:
But I changed it anyhow.